File: pkg/runtime/runtimes/python/pulserpc/validation.py
"""Validation functions for PulseRPC types"""

from typing import Any, Callable, Dict, List

from .rpctypes import find_struct, find_enum, get_struct_fields, ValidationError


def _join_path(parent: str, child: str) -> str:
    return f"{parent}.{child}" if parent else f".{child}"


def _array_index_path(parent: str, index: Any) -> str:
    return f"{parent}[{index}]"


def _make_error(path: str, message: str) -> ValidationError:
    return ValidationError(path=path, message=message)


def validate_string(value: Any, path: str = "") -> List[ValidationError]:
    if not isinstance(value, str):
        return [_make_error(path, f"Expected string, got {type(value).__name__}")]
    return []


def validate_int(value: Any, path: str = "") -> List[ValidationError]:
    if isinstance(value, bool):
        return [_make_error(path, f"Expected int, got bool")]
    if isinstance(value, int):
        return []
    if isinstance(value, float):
        if value != value:  # NaN check
            return [_make_error(path, f"Expected int, got float NaN")]
        if value == float('inf') or value == float('-inf'):
            return [_make_error(path, f"Expected int, got float infinity")]
        if value == int(value):
            return []
    return [_make_error(path, f"Expected int, got {type(value).__name__}")]


def validate_float(value: Any, path: str = "") -> List[ValidationError]:
    if not isinstance(value, (int, float)):
        return [_make_error(path, f"Expected float, got {type(value).__name__}")]
    return []


def validate_bool(value: Any, path: str = "") -> List[ValidationError]:
    if not isinstance(value, bool):
        return [_make_error(path, f"Expected bool, got {type(value).__name__}")]
    return []


def validate_array(
    value: Any,
    element_validator: Callable[[Any, str], List[ValidationError]],
    path: str = ""
) -> List[ValidationError]:
    if not isinstance(value, list):
        return [_make_error(path, f"Expected list, got {type(value).__name__}")]
    errors: List[ValidationError] = []
    for i, elem in enumerate(value):
        element_path = _array_index_path(path, i)
        errors.extend(element_validator(elem, element_path))
    return errors


def validate_map(
    value: Any,
    value_validator: Callable[[Any, str], List[ValidationError]],
    path: str = ""
) -> List[ValidationError]:
    if not isinstance(value, dict):
        return [_make_error(path, f"Expected dict, got {type(value).__name__}")]
    errors: List[ValidationError] = []
    for key, val in value.items():
        if not isinstance(key, str):
            errors.append(_make_error(path, f"Map key must be string, got {type(key).__name__}"))
            continue
        key_path = _array_index_path(path, key)
        errors.extend(value_validator(val, key_path))
    return errors


def validate_enum(
    value: Any,
    enum_name: str,
    allowed_values: List[str],
    path: str = ""
) -> List[ValidationError]:
    if not isinstance(value, str):
        return [_make_error(path, f"Expected string for enum {enum_name}, got {type(value).__name__}")]
    if value not in allowed_values:
        return [_make_error(path, f"Invalid value for enum {enum_name}: '{value}'. Allowed values: {allowed_values}")]
    return []


def validate_struct(
    value: Any,
    struct_name: str,
    struct_def: Dict[str, Any],
    all_structs: Dict[str, Any],
    all_enums: Dict[str, Any],
    path: str = ""
) -> List[ValidationError]:
    if not isinstance(value, dict):
        return [_make_error(path, f"Expected dict for struct {struct_name}, got {type(value).__name__}")]

    fields = get_struct_fields(struct_name, all_structs)
    errors: List[ValidationError] = []

    for field in fields:
        field_name = field['name']
        field_type = field['type']
        is_optional = field.get('optional', False)
        field_path = _join_path(path, field_name)

        if field_name not in value:
            if not is_optional:
                errors.append(_make_error(
                    field_path,
                    f"Missing required field '{field_name}' in struct {struct_name}"
                ))
        else:
            field_value = value[field_name]
            if field_value is None:
                if not is_optional:
                    errors.append(_make_error(
                        field_path,
                        f"Field '{field_name}' in struct {struct_name} cannot be None"
                    ))
            else:
                errors.extend(
                    validate_type(field_value, field_type, all_structs, all_enums, is_optional, field_path)
                )

    return errors


def validate_type(
    value: Any,
    type_def: Dict[str, Any],
    all_structs: Dict[str, Any],
    all_enums: Dict[str, Any],
    is_optional: bool = False,
    path: str = ""
) -> List[ValidationError]:
    if value is None:
        if is_optional:
            return []
        else:
            return [_make_error(path, "Value cannot be None for non-optional type")]

    if type_def.get('builtIn') == 'string':
        return validate_string(value, path)
    elif type_def.get('builtIn') == 'int':
        return validate_int(value, path)
    elif type_def.get('builtIn') == 'float':
        return validate_float(value, path)
    elif type_def.get('builtIn') == 'bool':
        return validate_bool(value, path)
    elif type_def.get('array'):
        element_type = type_def['array']
        element_validator = lambda v, p: validate_type(v, element_type, all_structs, all_enums, False, p)
        return validate_array(value, element_validator, path)
    elif type_def.get('mapValue'):
        value_type = type_def['mapValue']
        value_validator = lambda v, p: validate_type(v, value_type, all_structs, all_enums, False, p)
        return validate_map(value, value_validator, path)
    elif type_def.get('userDefined'):
        user_type = type_def['userDefined']
        struct_def = find_struct(user_type, all_structs)
        if not struct_def and '.' not in user_type:
            for qualified_key in all_structs:
                if qualified_key.endswith('.' + user_type):
                    struct_def = all_structs[qualified_key]
                    user_type = qualified_key
                    break
        if struct_def:
            return validate_struct(value, user_type, struct_def, all_structs, all_enums, path)
        else:
            enum_def = find_enum(user_type, all_enums)
            if not enum_def and '.' not in user_type:
                for qualified_key in all_enums:
                    if qualified_key.endswith('.' + user_type):
                        enum_def = all_enums[qualified_key]
                        break
            if enum_def:
                allowed_values = [v['name'] for v in enum_def.get('values', [])]
                return validate_enum(value, user_type, allowed_values, path)
            else:
                return [_make_error(path, f"Unknown user-defined type: {user_type}")]
    else:
        return [_make_error(path, f"Invalid type definition: {type_def}")]

File: pkg/runtime/runtimes/ts/pulserpc/validation.ts
/**
 * Validation functions for PulseRPC types
 */

import { findStruct, findEnum, getStructFields, TypeDef, StructMap, EnumMap, StructDef } from "./types";

export function validateString(value: any): void {
  if (typeof value !== "string") {
    throw new TypeError(`Expected string, got ${typeof value}`);
  }
}

export function validateInt(value: any): void {
  // Validate that value is a number with no fractional component
  // Values like 5.0 should pass (effectively an integer), but 5.1 should fail
  if (typeof value !== "number") {
    throw new TypeError(`Expected number for int, got ${typeof value}`);
  }
  if (Number.isInteger(value)) {
    return;
  }
  // Accept whole-number floats (e.g., 5.0, -3.0)
  if (value === Math.floor(value)) {
    return;
  }
  throw new TypeError(`Expected integer, got number with fractional component: ${value}`);
}

export function validateFloat(value: any): void {
  if (typeof value !== "number") {
    throw new TypeError(`Expected number for float, got ${typeof value}`);
  }
}

export function validateBool(value: any): void {
  if (typeof value !== "boolean") {
    throw new TypeError(`Expected boolean, got ${typeof value}`);
  }
}

export function validateArray(
  value: any,
  elementValidator: (v: any) => void
): void {
  if (!Array.isArray(value)) {
    throw new TypeError(`Expected array, got ${typeof value}`);
  }
  for (let i = 0; i < value.length; i++) {
    try {
      elementValidator(value[i]);
    } catch (e: any) {
      throw new Error(
        `Array element at index ${i} validation failed: ${e.message}`
      );
    }
  }
}

export function validateMap(
  value: any,
  valueValidator: (v: any) => void
): void {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(`Expected object for map, got ${typeof value}`);
  }
  for (const [key, val] of Object.entries(value)) {
    if (typeof key !== "string") {
      throw new TypeError(`Map key must be string, got ${typeof key}`);
    }
    try {
      valueValidator(val);
    } catch (e: any) {
      throw new Error(
        `Map value for key '${key}' validation failed: ${e.message}`
      );
    }
  }
}

export function validateEnum(
  value: any,
  enumName: string,
  allowedValues: string[]
): void {
  if (typeof value !== "string") {
    throw new TypeError(
      `Expected string for enum ${enumName}, got ${typeof value}`
    );
  }
  if (!allowedValues.includes(value)) {
    throw new Error(
      `Invalid value for enum ${enumName}: '${value}'. Allowed values: ${allowedValues.join(", ")}`
    );
  }
}

export function validateStruct(
  value: any,
  structName: string,
  structDef: any,
  allStructs: StructMap,
  allEnums: EnumMap
): void {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(
      `Expected object for struct ${structName}, got ${typeof value}`
    );
  }

  // Get all fields including parent fields
  const fields = getStructFields(structName, allStructs);

  // Check required fields
  for (const field of fields) {
    const fieldName = field.name;
    const fieldType = field.type;
    const isOptional = field.optional || false;

    if (!(fieldName in value)) {
      if (!isOptional) {
        throw new Error(
          `Missing required field '${fieldName}' in struct ${structName}`
        );
      }
    } else {
      // Field is present, validate it
      const fieldValue = value[fieldName];
      if (fieldValue === null) {
        if (!isOptional) {
          throw new Error(
            `Field '${fieldName}' in struct ${structName} cannot be null`
          );
        }
      } else if (fieldValue === undefined) {
        if (!isOptional) {
          throw new Error(
            `Field '${fieldName}' in struct ${structName} cannot be undefined`
          );
        }
      } else {
        // Create validator for this field type
        try {
          validateType(fieldValue, fieldType, allStructs, allEnums, isOptional);
        } catch (e: any) {
          throw new Error(
            `Field '${fieldName}' in struct ${structName} validation failed: ${e.message}`
          );
        }
      }
    }
  }
}

export function validateType(
  value: any,
  typeDef: TypeDef,
  allStructs: StructMap,
  allEnums: EnumMap,
  isOptional: boolean = false
): void {
  // Handle optional types - only null is valid, not undefined
  if (value === null) {
    if (isOptional) {
      return;
    } else {
      throw new Error("Value cannot be null for non-optional type");
    }
  }

  // undefined is never valid (not even for optional fields)
  if (value === undefined) {
    throw new Error("Value cannot be undefined");
  }

  // Built-in types
  if (typeDef.builtIn === "string") {
    validateString(value);
  } else if (typeDef.builtIn === "int") {
    validateInt(value);
  } else if (typeDef.builtIn === "float") {
    validateFloat(value);
  } else if (typeDef.builtIn === "bool") {
    validateBool(value);
  }
  // Array types
  else if (typeDef.array) {
    const elementType = typeDef.array;
    const elementValidator = (v: any) =>
      validateType(v, elementType, allStructs, allEnums, false);
    validateArray(value, elementValidator);
  }
  // Map types
  else if (typeDef.mapValue) {
    const valueType = typeDef.mapValue;
    const valueValidator = (v: any) =>
      validateType(v, valueType, allStructs, allEnums, false);
    validateMap(value, valueValidator);
  }
  // User-defined types
  else if (typeDef.userDefined) {
    const userType = typeDef.userDefined;
    // Check if it's a struct
    const structDef = findStruct(userType, allStructs);
    if (structDef) {
      validateStruct(value, userType, structDef, allStructs, allEnums);
    }
    // Check if it's an enum
    else {
      const enumDef = findEnum(userType, allEnums);
      if (enumDef) {
        const allowedValues = enumDef.values.map((v) => v.name);
        validateEnum(value, userType, allowedValues);
      } else {
        throw new Error(`Unknown user-defined type: ${userType}`);
      }
    }
  } else {
    throw new Error(`Invalid type definition: ${JSON.stringify(typeDef)}`);
  }
}

